lib/trino-parquet/src/main/java/io/trino/parquet/writer/ParquetMetadataUtils.java

```diff
@@ -0,0 +1,219 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.parquet.writer;
+
+import org.apache.parquet.column.statistics.BinaryStatistics;
+import org.apache.parquet.format.Statistics;
+import org.apache.parquet.internal.column.columnindex.BinaryTruncator;
+import org.apache.parquet.io.api.Binary;
+import org.apache.parquet.schema.ColumnOrder;
+import org.apache.parquet.schema.LogicalTypeAnnotation;
+import org.apache.parquet.schema.PrimitiveType;
+
+import java.util.Arrays;
+import java.util.Optional;
+
+// Based on org.apache.parquet.format.converter.ParquetMetadataConverter
+public class ParquetMetadataUtils
+{
+    private ParquetMetadataUtils() {}
+
+    public static <T extends Comparable<T>> Statistics toParquetStatistics(org.apache.parquet.column.statistics.Statistics<T> stats, int truncateLength)
+    {
+        Statistics formatStats = new Statistics();
+        // We skip the `withinLimit(stats, truncateLength)` check from equivalent parquet-mr code
+        // which avoids writing stats altogether if they're above 4K in size.
+        // This is to allow readers to take advantage of truncated stats for filtering rather than having no stats.
```

Comment:
What could be the drawbacks? Will other engines (Spark) be able to read such files?

Reply:
> I don't think it is safe for readers to use min/max values for strings to optimize aggregations as regardless of the code here

Right, but could there be major reader that does exactly what @rdblue mentioned?

Reply:
Spark implements this aggregation pushdown but they skip it for binary type for the above reason https://issues.apache.org/jira/browse/SPARK-36645
PrestoDB also implements aggregation pushdown to parquet but it's disabled by default and there is a separate flag for aggregation pushdown on binary type which is also disabled by default.
As far as I can tell, it is well known that binary stats can be truncated and until parquet format adds a way to detect it, using this for aggregation is not safe.

Reply:
@raunaqmorarka, I think the main issue is that the spec doesn't allow for truncation in the statistics min/max values (it does for ColumnIndex, but not here).  So regardless of what existing implementations may be doing, it's probably best to either include the complete value or omit it.

Reply:
@danielcweeks I don't see anything in the spec that prohibits it. As far as I can tell, the spec doesn't define what is row group level min/max.
There is also this configurable `ParquetProperties#statisticsTruncateLength` in https://github.com/apache/parquet-mr/blob/9b5a962df3007009a227ef421600197531f970a5/parquet-column/src/main/java/org/apache/parquet/column/ParquetProperties.java#L105 which can be setup to something less than 4K to get truncated stats.
I don't understand why this exists in parquet-mr (which I assume is the reference implementation) if the spec has forbidden it somewhere. If it was expressly not allowed then Spark and PrestoDB wouldn't need checks preventing aggregation pushdown on strings.
Given the proliferation in parquet reader/writer implementations and the general looseness of parquet spec, it is far more important what the actual implementations are doing rather than what is written (or not) in the official spec.

Reply:
i agree with @raunaqmorarka except that I value the spec higher. I love the spec-first approach.
If the real world departs from the spec significantly, then however, it's too late and spec needs to be updated.

For varchars we have truncation (i hope it's not just truncation for max values, @raunaqmorarka ? that would be incorrect). For timestamps we have rounding to millis. The min/max values in the manifest need to be take as lower/upper bounds, and data minimum and maximum value present in the data set.

Reply:
@raunaqmorarka I think the spec is pretty clear in stating that the `min_value` and `max_value` are the actual values (it shouldn't have to explicitly prohibit other interpretations).  We shouldn't ignore that in favor of existing implementations that may not adhere to it.

If you want to change that interpretation, then propose a change to the spec, but I would strongly discourage just ignoring it.

Reply:
@danielcweeks I didn't realize that comments in thrift file are also part of spec, I was looking at the md files for clarification.
I still think the `Min and max values for the column` at https://github.com/apache/parquet-format/blob/master/src/main/thrift/parquet.thrift#L219 could use an explicit comment to forbid string truncation given that parquet-mr has a string stats truncation feature.
Would you agree that the parquet-mr feature at https://github.com/apache/parquet-mr/blob/9b5a962df3007009a227ef421600197531f970a5/parquet-column/src/main/java/org/apache/parquet/column/ParquetProperties.java#L105 is already out of compliance with parquet spec ? 
Even if the parquet spec actually meant to disallow strings truncation, it still wouldn't make it safe for engines to implement aggregation pushdown on strings given that possibility of truncation exists and there is no way for the reader to detect it. Even existing attempts at aggregation pushdown into parquet in Spark and PrestoDB are already written with the assumption that string stats truncation is possible.
https://issues.apache.org/jira/browse/PARQUET-1685 introduced this truncation to parquet-mr.
It says in the design doc https://docs.google.com/document/d/1Mgb0dDXQJkgjouboDrGa9v06hWGJ0oPiwnmffXShQ_M/edit
```
If there is Parquet application requires the min/max value to be the real column value, then it will be broken. Given this feature is off by default and turning on it will require testing, it should be acceptable. 
```
Even if the feature is behind a flag, it still breaks previous expectations about truncation when it is configured and there is no way to detect whether or not truncation took place. I find it hard to take the spec seriously when the reference implementation finds it acceptable to break it behind a config flag.

Reply:
> as we are implementing Iceberg aggregation push down for ICebergConnector, and for min/max depends on the `lower_bound` and `upper_bound`, and count depends on `value_counts`

`value_counts` can be trusted, since this is the exact record count
`lower_bound` can be trusted to be a lower bound, so cannot be taken in lieu of `min(col)`

Reply:
@raunaqmorarka string truncation is important to be allowed, because if eg max value happens to be hundreds kb long, you don't want to keep track of that exactly in file footer and iceberg manifests (that's why Iceberg manifest requires upper_bound to be a bound, not exact maximum value).

I agree with @danielcweeks that ignoring spec and being non-conformant is not a desirable situation. I think we should offer a PR towards Parquet spec to align it with reality.

Reply:
I've filed https://issues.apache.org/jira/browse/PARQUET-2352 for the above to start a discussion about it in parquet

Reply:
Just for clarification:  Parquet started with `min`/`max` value because the issues with storing large values in stats wasn't initially apparent.  Hence the workaround of dropping the large values (as opposed to truncating).  @raunaqmorarka I would say that the current truncation feature puts the implementation out of spec if used.

When we designed Iceberg, we specifically used `lower`/`upper` bounds to avoid the issues discovered with the Parquet approach.  For Iceberg stats, you should never treat the bounds as the actual values for pushdown (they might be the min/max from the parquet stats, but we actively truncate for size/partitioning/etc., so you should never expect it to be the exact bound).

I'm all for updating the parquet spec if we think it's necessary, but I could also see the argument that it's a bit of an edge case and possible best to just drop like the current approach.

Reply:
I don't like departing from spec, but thinking about this pragmatically, following the spec faithfully buys us nothing. Every implementor of query engine on top of Parquet files need to assume min/max values are not necessarily exact, because widely used software writes them inexact. Thus, by writing truncated min/max we don't do more damage to the Parquet community, the damage already has been done.  Writing truncated min/max has the benefits compared with lack of stats -- it enables pruning, and we have real-life use-cases where some values are long, but still the column can usefully be searched over.

Thus, I think we should do everything we can to bring Parquet spec to reality.
Regardless whether this succeeds, we should write truncated min/max for long columns.

Reply:
Created PR https://github.com/apache/parquet-format/pull/216 for the parquet spec